src/main/java/net/sf/jabref/model/database/DuplicationChecker.java

```diff
@@ -96,17 +71,35 @@ protected boolean addKeyToSet(String key) {
      * <br>
      * Special case: If a null or empty key is passed, it is not counted and thus not removed.
      */
-    protected void removeKeyFromSet(String key) {
-        if ((key == null) || key.isEmpty()) {
+    private void removeKeyFromSet(String key) {
+        if (key == null || key.isEmpty()) {
             return;
         }
-        if (allKeys.containsKey(key)) {
-            Integer tI = allKeys.get(key);
-            if (tI == 1) {
-                allKeys.remove(key);
-            } else {
-                allKeys.put(key, tI - 1);
-            }
+
+        int numberOfKeyOccurrences = getNumberOfKeyOccurrences(key);
+        if (numberOfKeyOccurrences > 1) {
+            allKeys.put(key, numberOfKeyOccurrences - 1);
+        } else {
+            allKeys.remove(key);
+        }
+    }
+
+    @Subscribe
+    public void listen(FieldChangedEvent fieldChangedEvent) {
+        if (fieldChangedEvent.getFieldName().equals(BibEntry.KEY_FIELD)) {
+            removeKeyFromSet(fieldChangedEvent.getOldValue());
+            addKeyToSet(fieldChangedEvent.getNewValue());
         }
     }
+
+    @Subscribe
+    public void listen(EntryRemovedEvent entryRemovedEvent) {
+        removeKeyFromSet(entryRemovedEvent.getBibEntry().getCiteKeyOptional().orElse(null));
```

Comment:
if the key is `null` `removeKeyFromSet` will return immediately.
now I use the `ifPresent` method of the optional.